src/java_bytecode/generate_java_generic_method.cpp

```diff
@@ -0,0 +1,296 @@
+  java_generic_parametert &generic_parameter,
+  const generate_java_generic_methodt::type_variable_instantiation_mapt
+    &concrete_types) const
+{
+  INVARIANT(generic_parameter.id()==ID_pointer,
+            "All generic parameters should be pointers in java");
+  INVARIANT(generic_parameter.subtype().id()==ID_symbol,
+            "All generic parameters should point to symbols");
+
+  const symbol_typet &generic_type_variable=generic_parameter
+    .type_variable();
+  const auto &instantiation_type=concrete_types.find
+    (generic_type_variable);
+  if(instantiation_type==concrete_types.end())
+  {
+    // If we ever want/need to support partially instantiated types,
+    // we'll probably want to just return here.
+    throw
+      "No concrete type supplied for generic type variable in parameter";
+  }
+  generic_parameter.subtype()=instantiation_type->second;
+  generic_parameter.remove(ID_C_java_generic_parameter);
+  generic_parameter.remove(ID_type_variables);
+}
+
+
+/// Instantiate a given java generic type using a supplied mapping from type
+/// variable to concrete type.
+/// \param generic_type The java_generic_typet that should be modified and
+/// instantiated.
+/// \param concrete_types The mapping of type variables to concrete types.
+const void generate_java_generic_methodt::instantiate_java_generic_type(
+  java_generic_typet &generic_type,
+  const generate_java_generic_methodt::type_variable_instantiation_mapt
+    &concrete_types) const
+{
+  INVARIANT(generic_type.id()==ID_pointer,
+            "All generic parameters should be pointers in java");
+  INVARIANT(generic_type.subtype().id()==ID_symbol,
+            "All generic parameters should point to symbols");
+
+  // Replace the generic parameters
+  for(java_generic_parametert &generic_parameter :
+    to_java_generic_type(generic_type).generic_type_variables())
+  {
+    if(!is_java_generic_inst_parameter(generic_parameter))
+    {
+      const symbol_typet &generic_type_variable=generic_parameter
+        .type_variable();
+      const auto &instantiation_type=concrete_types.find
+        (generic_type_variable);
+      if(instantiation_type==concrete_types.end())
+      {
+        // If we ever want/need to support partially instantiated types,
+        // we'll probably want to just return here.
+        throw "no concrete type supplied for generic type parameter";
+      }
+      generic_parameter.subtype()=instantiation_type->second;
+      generic_parameter.set(ID_C_java_generic_inst_parameter, true);
+    }
+  }
+  // FIXME: At this point we have a concrete types substituted into the
+  // FIXME: generic type variables, but it's still a generic type.
+  // FIXME: I think once the rest of the generics stuff is committed,
+  // FIXME: particularly, once we have specialization of classes working,
+  // FIXME: we'll want to then trigger and/or replace the whole generic type
+  // FIXME: expression here with the concrete type.
```

Comment:
I suppose the intention here is to fix this before merging (otherwise I suggest raising a ticket on JIRA to track it)

Reply:
Yeah, I was intended that we'd updated this prior to merge.